core/src/avm2/object/xml_list_object.rs

```diff
@@ -230,6 +329,66 @@ impl<'a, 'gc> Deref for E4XWrapper<'a, 'gc> {
     }
 }
 
+/// Represents either a XmlObject or a XmlListObject. Used
+/// for resolving the value of empty XMLLists.
+#[derive(Clone, Collect, Copy, Debug)]
+#[collect(no_drop)]
+pub enum XmlOrXmlListObject<'gc> {
+    XmlList(XmlListObject<'gc>),
+    Xml(XmlObject<'gc>),
+}
+
+impl<'gc> XmlOrXmlListObject<'gc> {
+    pub fn length(&self) -> Option<usize> {
+        match self {
+            XmlOrXmlListObject::Xml(x) => x.length(),
+            XmlOrXmlListObject::XmlList(x) => Some(x.length()),
+        }
+    }
+
+    pub fn resolve_value(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+    ) -> Result<Option<XmlOrXmlListObject<'gc>>, Error<'gc>> {
+        match self {
+            // NOTE: XmlObjects just resolve to themselves.
+            XmlOrXmlListObject::Xml(x) => Ok(Some(XmlOrXmlListObject::Xml(*x))),
+            XmlOrXmlListObject::XmlList(x) => x.resolve_value(activation),
+        }
+    }
+
+    pub fn as_object(&self) -> Object<'gc> {
+        match self {
+            XmlOrXmlListObject::Xml(x) => Object::XmlObject(*x),
+            XmlOrXmlListObject::XmlList(x) => Object::XmlListObject(*x),
+        }
+    }
+
+    pub fn from_value(value: Value<'gc>) -> Result<XmlOrXmlListObject<'gc>, Error<'gc>> {
+        if let Some(xml) = value.as_object().and_then(|x| x.as_xml_object()) {
+            return Ok(XmlOrXmlListObject::Xml(xml));
+        }
+
+        if let Some(list) = value.as_object().and_then(|x| x.as_xml_list_object()) {
+            return Ok(XmlOrXmlListObject::XmlList(list));
+        }
+
+        Err(format!("Value {:?} was neither XmlObject or XmlListObject", value).into())
```

Comment:
If this error condition can never be hit, can you replace this with a `panic!`?

Reply:
This error condition technically should never be hit since get_property_local of Xml/XmlListObject should return a XmlListObject.

Reply:
In that case can you replace it with a panic?

Reply:
> In that case can you replace it with a panic?

I think so.

Reply:
> since get_property_local of Xml/XmlListObject should return a XmlListObject.

Sure, but this only applies to its callers, not to the method itself. I'd say adding such an invisible invariant to the signature is a bit weird. Maybe wrap the entire get-property + cast into a single method?